### PR TITLE
rubocop: lib/jekyll/renderer.rb complexity fixes

### LINKED ISSUE #4885: Make jekyll/jekyll code compliant with rubocop rules
Rubocop can help us keep our code consistent in style. Consistency makes it easier to read the code and having Rubocop handle reporting offenses makes feedback to pull requests quick & less controversial.

Instructions for fixing a file:
1. Ensure there isn't an existing pull request fixing the files you're looking at updating
2. Remove the file from `.rubocop.yml` under `AllCops.Exclude`.
3. Run `bundle install` to ensure you have all the gems.
4. Run `script/fmt`.
5. Fix errors until `script/fmt` reports no errors.
6. Commit to a feature branch on your fork and create a pull request mentioning this issue.
7. Rejoice!

Here's a checklist of files **in-progress**:
- [x] Fix offenses in `lib/jekyll/convertible.rb` and enable in `.rubocop.yml` (#5031, #5308)
- [x] Fix offenses in `lib/jekyll/document.rb` and enable in `.rubocop.yml` (#5030, #5045)
- [x] Fix offenses in `lib/jekyll/collection.rb` and enable in `.rubocop.yml` (#5022)
- [x] Fix offenses in `lib/jekyll/configuration.rb` and enable in `.rubocop.yml` (#5024)
- [x] Fix offenses in `lib/jekyll/regenerator.rb` and enable in `.rubocop.yml` (#5025)
- [ ] Fix offenses in `lib/jekyll/renderer.rb` and enable in `.rubocop.yml` (#5052)
- [x] Fix offenses in `lib/jekyll/utils.rb` and enable in `.rubocop.yml` (#5026)

Here's a checklist of completed tasks:
- [x] Finalize Rubocop rules (#4886)
- [x] Enable Rubocop check for pull requests (#4886)
- [x] Fix offenses in `exe/jekyll` and enable in `.rubocop.yml` (#5017)
- [x] Fix offenses in `lib/jekyll/cleaner.rb` and enable in `.rubocop.yml` (#4892)
- [x] Fix offenses in `lib/jekyll/commands/build.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/commands/clean.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/commands/doctor.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/commands/help.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/commands/new.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/commands/serve/servlet.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/commands/serve.rb` and enable in `.rubocop.yml` (#4888)
- [x] Fix offenses in `lib/jekyll/command.rb` and enable in `.rubocop.yml` (#5018)
- [x] Fix offenses in `lib/jekyll/converter.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/identity.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/markdown` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/markdown/kramdown_parser.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/markdown/rdiscount_parser.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/markdown/redcarpet_parser.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/markdown.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/converters/smartypants.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/deprecator.rb` and enable in `.rubocop.yml` (#4940)
- [x] Fix offenses in `lib/jekyll/drops/collection_drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/drops/document_drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/drops/drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/drops/jekyll_drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/drops/site_drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/drops/unified_payload_drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/drops/url_drop.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/entry_filter.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/errors.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/excerpt.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/external.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/filters.rb` and enable in `.rubocop.yml` (#4993)
- [x] Fix offenses in `lib/jekyll/frontmatter_defaults.rb` and enable in `.rubocop.yml` (#4974)
- [x] Fix offenses in `lib/jekyll/generator.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/hooks.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/layout.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/liquid_extensions.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/liquid_renderer/file.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/liquid_renderer/table.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/liquid_renderer.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/log_adapter.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/page.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/plugin.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/plugin_manager.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/publisher.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/readers/collection_reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/readers/data_reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/readers/layout_reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/readers/page_reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/readers/post_reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/readers/static_file_reader.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/related_posts.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/renderer.rb` and enable in `.rubocop.yml` (#4301)
- [x] Fix offenses in `lib/jekyll/site.rb` and enable in `.rubocop.yml` (#4973)
- [x] Fix offenses in `lib/jekyll/static_file.rb` and enable in `.rubocop.yml` (#4301)
- [x] Fix offenses in `lib/jekyll/stevenson.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/tags/highlight.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/tags/include.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/tags/link.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/tags/post_url.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/theme.rb` and enable in `.rubocop.yml` (#4959)
- [x] Fix offenses in `lib/jekyll/url.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `lib/jekyll/utils/ansi.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `lib/jekyll/utils/platforms.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `lib/jekyll/version.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `lib/jekyll.rb` and enable in `.rubocop.yml` (#4966)
- [x] Fix offenses in `features/step_definitions.rb` and enable in `.rubocop.yml` (#4956)
- [x] Fix offenses in `features/support/formatter.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `features/support/helpers.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/helper.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/simplecov_custom_profile.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_ansi.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_cleaner.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_coffeescript.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_collections.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_command.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_commands_serve.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_configuration.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_convertible.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_doctor_command.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_document.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_entry_filter.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_excerpt.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_filters.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_front_matter_defaults.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_generated_site.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_kramdown.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_layout_reader.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_liquid_extensions.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_liquid_renderer.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_log_adapter.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_new_command.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_page.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_path_sanitization.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_plugin_manager.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_rdiscount.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_redcarpet.rb` and enable in `.rubocop.yml` (#4902)
- [x] Fix offenses in `test/test_regenerator.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_related_posts.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `test/test_sass.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `test/test_site.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_static_file.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `test/test_tags.rb` and enable in `.rubocop.yml`
- [x] Fix offenses in `test/test_theme.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `test/test_url.rb` and enable in `.rubocop.yml` (#4916)
- [x] Fix offenses in `test/test_utils.rb` and enable in `.rubocop.yml`

### PR DESCRIPTION
**Pull request by ayastreb**
#4885

In this PR I refactored `Renderer` class, removed duplicated code from `Convertible` and changed `Page` class to use `Renderer.run` instead of `Convertible.do_layout`

### CODE REVIEW

File: lib/jekyll/renderer.rb [hunk 1]
```diff
@@ -13,78 +13,96 @@ def initialize(site, document, site_payload = nil)
...
-      }
-
-      output = document.content
-
+    # Render the document.
+    # Puts rendered content into document.output
+    #
+    # Returns nothing
+    private
+    # rubocop: disable AbcSize
```
Comment (ayastreb) [inline lib/jekyll/renderer.rb]: The ABC metric is 25 in this method (20 is max).
I couldn't think of a way to reduce it without creating too many helper methods.
Any suggestions? 😄

**Comment by ayastreb** [comment]
I've rebased this PR to latest master and fixed conflicts.

File: lib/jekyll/page.rb [hunk 2]
```diff
@@ -123,12 +123,10 @@ def process(name)
...
     #
-    # Returns nothing.
+    # Returns String rendered page.
     def render(layouts, site_payload)
-      site_payload["page"] = to_liquid
-      site_payload["paginator"] = pager.to_liquid
-
-      do_layout(site_payload, layouts)
+      site.layouts = layouts
+      Renderer.new(site, self, site_payload).run
```
Comment (ayastreb) [inline lib/jekyll/page.rb]: @benbalter is doing similar change in this method [here](https://github.com/jekyll/jekyll/pull/5211/files#diff-6257bc8ccdf5697787354024f8d97a78R77), but skipping `layouts` variable.
Should we do the same here?

File: lib/jekyll/page.rb [hunk 3]
```diff
@@ -123,12 +123,10 @@ def process(name)
...
     # site_payload - The site payload Hash.
     #
-    # Returns nothing.
+    # Returns String rendered page.
     def render(layouts, site_payload)
-      site_payload["page"] = to_liquid
-      site_payload["paginator"] = pager.to_liquid
-
-      do_layout(site_payload, layouts)
+      site.layouts = layouts
```
Comment (parkr) [inline lib/jekyll/page.rb]: This is a no-go. We don't want to reassign `site.layouts`.

File: lib/jekyll/renderer.rb [hunk 4]
```diff
@@ -13,78 +13,96 @@ def initialize(site, document, site_payload = nil)
     # Determine which converters to use based on this document's
     # extension.
     #
-    # Returns an array of Converter instances.
+    # Returns Array of Converter instances.
     def converters
-      @converters ||= site.converters.select { |c| c.matches(document.extname) }
+      @converters ||= site.converters.select { |c| c.matches(document.extname) }.sort
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: Might break things but technically a bug.

File: lib/jekyll/renderer.rb [hunk 5]
```diff
@@ -13,78 +13,96 @@ def initialize(site, document, site_payload = nil)
...
+    # Puts rendered content into document.output
+    #
+    # Returns nothing
+    private
+    # rubocop: disable AbcSize
+    def render_document
       if document.render_with_liquid?
         Jekyll.logger.debug "Rendering Liquid:", document.relative_path
-        output = render_liquid(output, payload, info, document.path)
+        document.output = render_liquid(document.content, document.path)
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: Assigning `document.output` should not be done here – we should not modify `document` inside the renderer; it is up to the caller to modify `document`.

File: lib/jekyll/convertible.rb [hunk 6]
```diff
@@ -72,24 +72,6 @@ def validate_permalink!(filename)
       end
     end
 
-    # Transform the contents based on the content type.
-    #
-    # Returns the transformed contents.
-    def transform
```
Comment (parkr) [inline lib/jekyll/convertible.rb]: We can't remove these from Convertible until Jekyll 4 (when we'll remove the whole module). These could proxy out to a `Jekyll::Renderer` instance created in a private helper method.

**Comment by parkr** [comment]
Thanks for this and sorry for the delay in reviewing. This appears to change too many things with the actual functionality and compatibility so I'm not going to merge it. Do you want to fix it up based on my comments? I could also accept a PR which just proxied the `Convertible` method calls to a `Jekyll::Renderer` instance then take on the renderer changes later.

File: lib/jekyll/page.rb (hunk 3 above)
Comment (ayastreb) [inline lib/jekyll/page.rb]: Ok, should we just ignore the `layouts` argument then?

File: lib/jekyll/renderer.rb (hunk 4 above)
Comment (ayastreb) [inline lib/jekyll/renderer.rb]: Hmm, that came from `Convertible` 😄 
What would you suggest to do?

**Comment by ayastreb** [comment]
@parkr 
Sure, I can open new PR after #5308 is merged and will only fix `Renderer` complexity issues there. Let's keep this PR open until then?

**Comment by parkr** [comment]
Merged! Would you mind rebasing?

**Comment by ayastreb** [comment]
@parkr done!

File: lib/jekyll/convertible.rb [hunk 7]
```diff
@@ -193,7 +193,7 @@ def invalid_layout?(layout)
     # Returns nothing
     def render_all_layouts(layouts, payload, info)
       _renderer.layouts = layouts
-      _renderer.place_in_layouts(output, payload, info)
+      _renderer.place_in_layouts(payload, info)
```
Comment (parkr) [inline lib/jekyll/convertible.rb]: We can't change any API signatures during a rubocop refactor. We don't know what plugins use this method and don't want to break them. Please revert this change!

**Comment by ayastreb** [comment]
@parkr ok, I've reverted all signature changes

File: lib/jekyll/page.rb [hunk 8]
```diff
@@ -123,12 +123,11 @@ def process(name)
...
     # site_payload - The site payload Hash.
     #
-    # Returns nothing.
+    # Returns String rendered page.
     def render(layouts, site_payload)
-      site_payload["page"] = to_liquid
-      site_payload["paginator"] = pager.to_liquid
-
-      do_layout(site_payload, layouts)
+      renderer = Renderer.new(site, self, site_payload)
```
Comment (parkr) [inline lib/jekyll/page.rb]: This class still includes `Convertible`. You should use the `_renderer` method here instead and just set payload and layouts. Like:

```ruby
self.output = _renderer.tap do |renderer|
  renderer.payload = site_payload
  renderer.layouts = layouts
end.run
@_renderer = nil
self.output
```

:)

Comment (parkr) [inline lib/jekyll/page.rb]: Ah, I did something like this in #5337.

File: lib/jekyll/renderer.rb [hunk 9]
```diff
@@ -33,76 +33,65 @@ def layouts
...
-      else
-        payload["site"]["related_posts"] = nil
-      end
-
-      # render and transform content (this becomes the final content of the object)
-      payload["highlighter_prefix"] = converters.first.highlighter_prefix
-      payload["highlighter_suffix"] = converters.first.highlighter_suffix
+      assign_pages
+      assign_related_posts
+      assign_highlighter_options
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: I'd love for these to each be "bang" methods because they modify the payload:

```ruby
assign_pages!
assign_related_posts!
assign_highlighter_options!
```

File: lib/jekyll/renderer.rb [hunk 10]
```diff
@@ -33,76 +33,65 @@ def layouts
...
+      render_document
+    end
+
+    # Render the document.
+    #
+    # Returns String rendered document output
+    # rubocop: disable AbcSize
+    def render_document
       info = {
+        :filters   => [Jekyll::Filters],
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: I think we removed this on purpose somewhere... Why did you add it back out of curiosity?

File: lib/jekyll/renderer.rb [hunk 11]
```diff
@@ -33,76 +33,65 @@ def layouts
...
         :registers => { :site => site, :page => payload["page"] }
       }
-
-      output = document.content
-
+      @output = document.content
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: I'd rather not keep this as an instance variable. Is it used outside the scope and not passed? I think just `output` would be better and to just pass it to the methods that need it. Less state for this class we have to reason about the better. 😄

File: lib/jekyll/renderer.rb [hunk 12]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
-    #
-    # content - the content to be placed in the layout
-    #
+    # Render layouts and place document content inside.
     #
-    # Returns the content placed in the Liquid-rendered layouts
+    # Returns nothing
     def place_in_layouts(content, payload, info)
-      output = content.dup
+      @output = content.dup
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: Same here – just `output` would be better in my opinion.

File: lib/jekyll/renderer.rb [hunk 13]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
-
-        # Add layout to dependency tree
-        site.regenerator.add_dependency(
-          site.in_source_dir(document.path),
-          site.in_source_dir(layout.path)
-        ) if document.write?
-
-        if (layout = layouts[layout.data["layout"]])
+        render_layout(layout, info)
+        add_layout_to_dependency(layout)
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: Let's rephrase this method – I was a bit confused by its name. What do you think about `add_regenerator_dependencies(layout)`?

File: lib/jekyll/renderer.rb [hunk 14]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
+        "in #{document.relative_path} does not exist."
+      ) if invalid_layout?(layout)
+    end
+
+    # Render layout content into document.output
+    #
+    # Returns nothing
+    private
+    def render_layout(layout, info)
+      payload["content"] = @output
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: I think passing the value of output as an argument would be easier to reason about.

File: lib/jekyll/renderer.rb [hunk 15]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
+      payload["layout"]  = Utils.deep_merge_hashes(layout.data, payload["layout"] || {})
+
+      @output = render_liquid(
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: And then returning it here instead of using the instance variable.

File: lib/jekyll/renderer.rb [hunk 16]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
+        site.in_source_dir(document.path),
+        site.in_source_dir(layout.path)
+      ) if document.write?
+    end
+
+    # Set page content to payload and assign pager if document has one.
+    #
+    # Returns nothing
+    private
+    def assign_pages
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: `assign_pages!`

File: lib/jekyll/renderer.rb [hunk 17]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
+        payload["paginator"] = document.pager.to_liquid
+      end
+    end
 
-      output
+    # Set related posts to payload if document is a post.
+    #
+    # Returns nothing
+    private
+    def assign_related_posts
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: `assign_related_posts!`

File: lib/jekyll/renderer.rb [hunk 18]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
+    end
+
+    # Set page content to payload and assign pager if document has one.
+    #
+    # Returns nothing
+    private
+    def assign_pages
+      payload["page"] = document.to_liquid
+      if document.respond_to?(:pager)
+        payload["paginator"] = document.pager.to_liquid
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: Let's set the `paginator` to `nil` if the document doesn't response, just like we do in the similar situation of `assign_related_posts!`.

File: lib/jekyll/renderer.rb [hunk 19]
```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
...
+      else
+        payload["site"]["related_posts"] = nil
+      end
     end
 
+    # Set highlighter prefix and suffix
+    #
+    # Returns nothing
     private
+    def assign_highlighter_options
```
Comment (parkr) [inline lib/jekyll/renderer.rb]: `assign_highlighter_options!`

**Comment by parkr** [comment]
Whew! OK, I left all my comments. This is looking pretty good! Let me know if you don't want to make these edits and I can merge your PR locally, make the edits, and push to master. 👍

**Comment by parkr** [comment]
Also: thank you! ❤️

**Comment by ayastreb** [comment]
Hey @parkr ,
thanks for the feedback! I will try to do the edits tomorrow 😄

File: lib/jekyll/renderer.rb (hunk 10 above)
Comment (ayastreb) [inline lib/jekyll/renderer.rb]: I think I copied it over from `Convertible`, it was used there -https://github.com/jekyll/jekyll/pull/5308/files#diff-817b22aa9906612f8c43543d6aa8202dL262

should it be removed? why was it used in `Convertible`? 😄

File: lib/jekyll/renderer.rb (hunk 11 above)
Comment (ayastreb) [inline lib/jekyll/renderer.rb]: Hmm... I think it made more sense before when I changed the signatures.
But since we can't change the signatures, I'll roll it back to just pass it to the methods!

File: lib/jekyll/page.rb [hunk 20]
```diff
@@ -123,12 +123,12 @@ def process(name)
...
+    # Returns String rendered page.
     def render(layouts, site_payload)
-      site_payload["page"] = to_liquid
-      site_payload["paginator"] = pager.to_liquid
-
-      do_layout(site_payload, layouts)
+      self.output = _renderer.tap do |renderer|
+        renderer.layouts = layouts
+        renderer.payload = site_payload
+      end.run
```
Comment (ayastreb) [inline lib/jekyll/page.rb]: @parkr since we're not removing Convertible at least till Jekyll 4, maybe I should just rollback this change and just call `do_layout` as it was before?
Because now it just duplicates part of the `Convertible.do_layout`

**Comment by ayastreb** [comment]
Ok, I've done all the edits and rebased the branch, should be ready now 😄

File: lib/jekyll/page.rb (hunk 20 above)
Comment (parkr) [inline lib/jekyll/page.rb]: Yeah, I think that would work!

**Comment by pathawks** [comment]
@parkr Is this good to merge?

**Comment by parkr** [comment]
> @parkr Is this good to merge?

It looks like I have some unaddressed questions! I'd also love to see this brought up-to-date with `master`.

**Comment by parkr** [comment]
Merged in 1c33bd579796d6f77abafc1af78e496298b4d74d 🎉 !
